petastorm/pytorch.py

```diff
@@ -118,7 +147,10 @@ def __iter__(self):
         self._in_iter = True
 
         try:
-            for batch in self._iter_impl():
+            iterator = self._iter_impl()
+            if self._max_prefetch > 1:
+                iterator = BackgroundIterator(iterator, prefetch=self.max_prefetch)
```

Comment:
Will the `BackgroundIterator` be properly destroyed and the thread joined when we exit the function (either nominal exit or with an exception?).

Reply:
Need to check this, thank you for the heads up.

Reply:
I know it could be tricky, but would be good to also test these aspects in a unit test or two.